cmd/agent/dist/conf.d/containerd.d/conf.yaml.default

```diff
@@ -1,6 +1,7 @@
 ad_identifiers:
   - _containerd
 init_config:
+loader: core
```

Comment:
@djmitche Hmm I'm a bit conflicted on the implementation used here. While it makes sense and should probably work, I am a bit concerned about the use of a internal-ish permanently-static setting in files that users will be customizing. To me it seems more advantageous to UX to hardcode the loaders for specific check names that are only in `core` as their names are already known anyways (with an exclusion list for checks that use the same name in both `python` and `core` loaders). What are your thoughts on this?

Reply:
I suspect that the ideal fix here would be to have a way to probe for a Python check without logging errors.  This seems like a temporary workaround to that fix, but you're right -- this has some long-term consequences if we were ever to move one of these core checks to Python.  @olivielpeau what do you think?

Reply:
Users shouldn't customize `.yaml.default` files. These files are part of the package and get overwritten on agent upgrade.

If/when these core checks are moved to Python, we can update the related `.yaml.default` files at the same time. In this scenario, users who have already copied the `.yaml.default` file to a `.yaml` file would still load the `core` check, but that seems fine to me since these users are customizing the check's config anyway.

Reply:
> Users shouldn't customize .yaml.default files. These files are part of the package and get overwritten on agent upgrade.

Oh I know - the comment I made was for when someone copies these `foo.yaml.default` to `foo.yaml` which would retain `loader: core` which may not necessarily be wanted to be configurable/exposed as an option in terms of UX.

Reply:
> that seems fine to me

I read that to say this is OK as-is.  Sound good, @sgnn7 ?

Reply:
> when someone copies these foo.yaml.default to foo.yaml which would retain loader: core which may not necessarily be wanted to be configurable/exposed as an option in terms of UX.

You have a point, in general the `loader` should be an implementation details that users shouldn't care about, except in very specific scenarios (such as the SNMP check, which has a core and a python implementation that are slightly different and documented). Also, having the `loader` defined in `.yaml` files makes the migration from a `core` to a `python` check (or the other way around) difficult.

Overall, we may be trying to solve a problem for which the solutions are worse than the problem. Hardcoding a list of `core`-exclusive check names in the Agent adds more complexity to the loading logic of checks (and doesn't solve the problem for the checks that have both `core` and `python` implementations, i.e. `disk` and `network` checks) so I don't find this solution appealing either.

Are the debug-level errors misleading enough that it's worth trying to add such complexity to solve the problem?

Reply:
> Overall, we may be trying to solve a problem for which the solutions are worse than the problem. 

That may be the case

> Hardcoding a list of `core`-exclusive check names in the Agent adds more complexity to the loading logic of checks (and doesn't solve the problem for the checks that have both `core` and `python` implementations, i.e. `disk` and `network` checks) so I don't find this solution appealing either.

I think this may be due to the way we do check loading. Since the internal check list is static already I would think that the enumeration could be easy and I do wonder why we do not check `core` first for a match then `python` if `loader` isn't specified rather than the other way around (but I may not have some context around this that I should).

> Are the debug-level errors misleading enough that it's worth trying to add such complexity to solve the problem?

From my point of view, a user that is not a developer of agent-core would have a real hard time discerning from those logged messages if: a check is failing to load, if the python loader is failing to load a check since it's a core check, if there's a python check package missing, or if a python import of a package is failing due to missing dependencies. To make things worse, one of those we consider to be completely normal (not a "real" problem) and expected. Because of this wonky UX, I'm heavily inclined towards us fixing this if we can but it is a debug message and maybe it's not that big of a deal and/or we can work around this improved error messages.

Reply:
> I do wonder why we do not check core first for a match then python if loader isn't specified rather than the other way around

In general, the reason for the current loader precedence is that it allows any `core` check (which is compiled in the Agent) to be overridden at load time by a `python` check (which is loaded at runtime). So in general this order of precedence provides more flexibility at runtime. That property isn't really used though AFAIK (except for the `disk` and `network` checks).

> a check is failing to load, if the python loader is failing to load a check since it's a core check, if there's a python check package missing, or if a python import of a package is failing due to missing dependencies

This is difficult to improve because in all of these cases I believe Python returns an `ImportError`, so the Agent would have to do some introspection on the error's specifics to then return a more meaningful error. I think we already log the full error from python.

So overall, if we want to fix the issue properly, we should probably look into better handling the Python errors at load time instead of this PR. This would have its own challenges and would only fix what I see as a minor problem so I don't think we should prioritize it now. Happy to hear your thoughts though.

Reply:
> So overall, if we want to fix the issue properly, we should probably look into better handling the Python errors at load time instead of this PR. This would have its own challenges and would only fix what I see as a minor problem so I don't think we should prioritize it now. Happy to hear your thoughts though.

Yeah sounds fair to me